gsw_xarray/tests/test_imports.py

```diff
@@ -0,0 +1,142 @@
+    )
+
+
+gsw_base = get_module_names(gsw)
+wrapped_funcs = gsw_xarray._core._wrapped_funcs
+
+submodules = [
+    "conversions",
+    "density",
+    "energy",
+    "freezing",
+    "geostrophy",
+    "ice",
+    "stability",
+]
+
+
+@pytest.mark.parametrize("upstream_func", gsw_base)
+def test_wrapped_interface(upstream_func):
+    """Tests the complete gsw function namespace against ours"""
+    assert hasattr(gsw_xarray, upstream_func)
+
+
+@pytest.mark.parametrize("upstream_func", gsw_base)
+def test_wrapped_return(upstream_func):
+    if upstream_func in wrapped_funcs:
+        assert getattr(gsw_xarray, upstream_func) is wrapped_funcs[upstream_func]
+    else:
+        assert getattr(gsw_xarray, upstream_func) is getattr(gsw, upstream_func)
+
+
+@pytest.mark.parametrize("submodule", submodules)
+def test_wrapped_interface_submodules(submodule):
+    upstream_submodule = getattr(gsw, submodule)
+    wrapepd_submodule = getattr(gsw_xarray, submodule)
+
+    upstream_funcs = get_module_names(upstream_submodule)
+    for name in upstream_funcs:
+        if name in wrapped_funcs:
+            assert getattr(wrapepd_submodule, name) is wrapped_funcs[name]
+        else:
+            assert getattr(wrapepd_submodule, name) is getattr(upstream_submodule, name)
+
+
+@pytest.mark.parametrize("upstream_func", gsw_base)
+def test_import_interface_base(upstream_func):
+    """This tests the interface in the form of:
+    > from gsw_xarray import func
+    where func is any function on the upstream gsw
+
+    It does this by creating an empty "locals" dict and calling an exec on the way the code
+    would be written, it then checks the idenity of that returned result
+    """
+    result = {}
+    # the empty dict is for the "globals" that we don't care about (exec would populate it with builtins)
+    exec(f"from gsw_xarray import {upstream_func}", {}, result)
+    func = result[upstream_func]
+    if upstream_func in wrapped_funcs:
+        assert func is wrapped_funcs[upstream_func]
+    else:
+        assert func is getattr(gsw, upstream_func)
+
+
+@pytest.mark.parametrize("upstream_func", gsw_base)
+@pytest.mark.parametrize("submodule", submodules)
+def test_import_interface_submodule(submodule, upstream_func):
+    """This tests the interface in the form of:
+    > from gsw_xarray.submod import func
+    where submod is one of the submodule groupings in the upstream
+
+    """
+    result = {}
+    upstream_interface = get_module_names(getattr(gsw, submodule))
+
+    expected_func = getattr(gsw, upstream_func)
+
+    if upstream_func not in upstream_interface:
+        with pytest.raises(ImportError):
+            exec(f"from gsw_xarray.{submodule} import {upstream_func}", {}, result)
+    else:
+        exec(f"from gsw_xarray.{submodule} import {upstream_func}", {}, result)
+
+        func = result[upstream_func]
+
+        if isinstance(func, np.ufunc):
+            pytest.xfail(f"{func} of submodule {submodule} is a directly exposed ufunc")
```

Comment:
The upstream gsw doesn't directly expose the actual c compiled ufuncs, it wrapps them all in python funcs of the same name (with some decorators and doc strings). There are a few places in the upstream library that directly use the c ufuncs internally. Due to the way it's imported in those places, the name is exposed and importable.

So you are correct in that `gsw.stability.grav` is not the same as `gsw.grav` in the upstream. This could potentially be solved by only looking at what is in `__all__` but that really only matters when doing a wildcard import.

Reply:
Basically... the upstream leaks its private api a little. Not knowing what to do for these cases, I just xfailed them. Maybe we can raise with the other gsw-python maintainers.